apps/consulting/posts/scaling-python-banking-edition.md

```diff
@@ -0,0 +1,677 @@
+sums, so you won’t end up adding the wrong numbers together. Pandas sums are also considerably
+faster than group-by operations. As another bonus, moving out of Dask
+dataframes would allow us to utilize multi-indexing, which helps cut down on the
+memory footprint of the dataframes. We found that on our
+dataset, summing our re-indexed dataframe was about 3 times faster
+than running a group-by operation on the equivalent concatenated
+dataframe.
+
+```python
+>>> %timeit concated_list_of_dfs.groupby(['path', 'Date']).sum()
+1.77 s ± 72.3 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
+```
+
+<img
+src="/posts/scaling-python-banking-edition/groupby_sum_diagram.png"
+alt="An image of a dataframe that shows the before and after of an groupby operation."
+/>
+
+```python
+>>> %timeit concated_list_of_dfs.groupby(['path', 'Date']).sum()
+1.77 s ± 72.3 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
+```
+
+<img
+src="/posts/scaling-python-banking-edition/df_list_sum.png"
+alt="An image of a dataframe that shows the before and after diagram for summing a list of dataframes."
+/>
+
+**_!!!! (What's the difference between the two figures above? Why are both of these figures included? If they're showing two different things, that difference needs to be explained. If they're both showing basically the same thing, then one figure can be removed. Also, the two code snippets look identical—should they be different?) !!!!_**
+
+While these performance improvements were promising, we were still having a bit of trouble with getting the dataframes to
+the right place at the right time. **_!! (What does this mean? What is the 'right place', and what is the 'right time'?) !!_**
+
+We also wanted our
+architecture to exploit any parallelization opportunities among
+the different aggregation groups. Ideally, we wanted to run an
+aggregation group on a single worker whenever feasible, to cut down on
+network traffic on the cluster. With that in mind we started looking
+at a cluster-of-clusters approach.
+
+### Cluster-of-Clusters
+
+We based our cluster-of-clusters implementation off of a [keynote at
+the 2021 dask
+summit][cluster of clusters keynote]. With
+this setup, you have a normal Dask distributed cluster with a
+scheduler and a set of workers. Instead of passing your work directly
+to the scheduler as normal, though, you pass delayed objects that
+instruct the worker to start a local cluster and also to build a
+specific Dask graph to be passed to that local cluster. Submitting work
+in this way gives the benefit of being able to encapsulate a work unit
+in order to reduce the overall network communication. However, at this
+point we were still using thread-pools to manage the aggregation on
+the local clusters, requiring the same cumbersome manual tuning as before. So, from here we decided
+to pivot away from the aggregation strategy to a different approach.
+
+<img
+src="/posts/scaling-python-banking-edition/cluster-of-clusters.png"
```

Comment:
Will most readers know the 'k8s' abbreviation? I think we should at least note that k8s = Kubernetes somewhere in the text. If there's no good place to put it in the flow of the narrative, then something like this below the figure would be fine.

```
&nbsp; &nbsp; _(k8s = Kubernetes)_
```

Reply:
I added a caption between the two images with k8s